### PR TITLE
feat: migrate webhook runner configuration to SSM

### PR DESCRIPTION
**Pull request by GuptaNavdeep1983**
This PR migrates the confugration for the webhook from environment variables to SSM to avoid the maximum size of environment variables is reached.

## Implementation

The webhook will read the configuration from SSM as json string. As long the lambda is hot the configuration is cached to speed-up the lambda time. In cases of configuration changes Lambda resources will be re-created by Terraform to ensure no cached values are used.

fix: #3594

### CODE REVIEW

**Comment by spcaipers-arm** [comment]
Hey guys, any idea when this is going to be merged?

**Comment by GuptaNavdeep1983** [comment]
> Hey guys, any idea when this is going to be merged?

There were few comments on this PR. Addressed and tested them today. Hopefully we should be able to merge this week.

**Review by npalm — COMMENTED** [review]
Still working on the review and test.

File: modules/webhook/variables.tf [hunk 1]
```diff
@@ -22,8 +22,8 @@ variable "tags" {
   default     = {}
 }
 
-variable "runner_config" {
-  description = "SQS queue to publish accepted build events based on the runner type. When exact match is disabled the webhook accecpts the event if one of the workflow job labels is part of the matcher. The priority defines the order the matchers are applied."
+variable "queues_config" {
```
[the remark this answers is not in the document]
Comment (GuptaNavdeep1983) [reply modules/webhook/variables.tf:25]: I have tried to explain the reason for it in the PR description.

File: modules/webhook/variables.tf [hunk 2]
```diff
@@ -35,7 +35,7 @@ variable "runner_config" {
     })
   }))
   validation {
-    condition     = try(var.runner_config.matcherConfig.priority, 999) >= 0 && try(var.runner_config.matcherConfig.priority, 999) < 1000
+    condition     = try(var.queues_config.matcherConfig.priority, 999) >= 0 && try(var.queues_config.matcherConfig.priority, 999) < 1000
```
Comment (npalm) [inline modules/webhook/variables.tf:38]: Should we not name this to runner or matcher config.

File: modules/webhook/policies/lambda-ssm.json [hunk 3]
```diff
@@ -3,11 +3,10 @@
...
     {
       "Effect": "Allow",
-      "Action": [
-        "ssm:GetParameter"
-      ],
+      "Action": ["ssm:GetParameter"],
       "Resource": [
-        "${github_app_webhook_secret_arn}"
+        "${github_app_webhook_secret_arn}",
+        "${parameter_queues_config_arn}"
```
Comment (npalm) [inline modules/webhook/policies/lambda-ssm.json:9]: Here also name it matcher / runner config

File: modules/webhook/main.tf [hunk 4]
```diff
@@ -55,3 +55,10 @@ resource "aws_apigatewayv2_integration" "webhook" {
   integration_method = "POST"
   integration_uri    = aws_lambda_function.webhook.invoke_arn
 }
+
+
+resource "aws_ssm_parameter" "runner_matcher_config" {
+  name  = "${var.ssm_paths.root}/runner-matcher-config"
```
Comment (npalm) [inline modules/webhook/main.tf:61]: Here include webhook in the path, so we can brinag later more webhook config in the same SSM path. Pererable via the object.

Ensure you also update the docs, see the onnfiguration docs (table with SSM paths).

**Review by npalm — CHANGES_REQUESTED** [review]
A few remarks

File: main.tf [hunk 5]
```diff
@@ -123,12 +123,14 @@ module "ssm" {
 
 module "webhook" {
   source = "./modules/webhook"
-
+  ssm_paths = {
+    root = "${local.ssm_root_path}"
```
Comment (npalm) [inline main.tf:127]: Root module was not working
```suggestion
    root = "${local.ssm_root_path}"
    webhook = "${var.ssm_paths.webhook}"
```

File: modules/webhook/webhook.tf [hunk 6]
```diff
@@ -50,6 +50,9 @@ resource "aws_lambda_function" "webhook" {
       mode = var.tracing_config.mode
     }
   }
+  lifecycle {
+    replace_triggered_by = [aws_ssm_parameter.runner_matcher_config]
```
Comment (npalm) [inline modules/webhook/webhook.tf:54]: Can we trigger also on change of the webhook secret?

File: lambdas/functions/webhook/src/ConfigResolver.ts [hunk 7]
```diff
@@ -1,15 +1,39 @@
...
+  constructor(repositoryAllowList: Array<string>, workflowJobEventSecondaryQueue: string | undefined) {
+    this.repositoryAllowList = repositoryAllowList;
+
+    this.workflowJobEventSecondaryQueue = workflowJobEventSecondaryQueue;
+  }
+
+  static async load(): Promise<Config> {
+    const repositoryAllowListEnv = process.env.REPOSITORY_ALLOW_LIST ?? '[]';
+    const repositoryAllowList = JSON.parse(repositoryAllowListEnv) as Array<string>;
+    // load the queues config from SSM if it's not already loaded and cached
```
Comment (npalm) [inline lambdas/functions/webhook/src/ConfigResolver.ts:20]: Comment can be something like, load parallel config if not cached

File: modules/webhook/webhook.tf [hunk 8]
```diff
@@ -65,6 +68,9 @@ resource "aws_lambda_permission" "webhook" {
   function_name = aws_lambda_function.webhook.function_name
   principal     = "apigateway.amazonaws.com"
   source_arn    = "${aws_apigatewayv2_api.webhook.execution_arn}/*/*/${local.webhook_endpoint}"
+  lifecycle {
+    replace_triggered_by = [aws_ssm_parameter.runner_matcher_config]
```
Comment (npalm) [inline modules/webhook/webhook.tf:72]: Why do we need the lifecyle for the loggroup?

**Review by npalm — COMMENTED** [review]
PR looks good, made a few minor comments. Can you also check the failing CI?

File: modules/webhook/webhook.tf (hunk 6 above)
Comment (GuptaNavdeep1983) [reply modules/webhook/webhook.tf]: done.

File: modules/webhook/webhook.tf (hunk 8 above)
Comment (GuptaNavdeep1983) [reply modules/webhook/webhook.tf]: This is not the log group but lambda permission resource which needs to be recreated every time lambda updates.

**Review by npalm — APPROVED** [review]

**Pull request merged**